src/cargo/ops/cargo_new.rs

```diff
@@ -44,40 +57,231 @@ struct CargoNewConfig {
+             target_name: project_name,
+             bin: true,
+        }
+    } else {
+        SourceFileInformation {
+             relative_path: String::from("src/lib.rs"),
+             target_name: project_name,
+             bin: false,
+        }
+    }
+}
+
+pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
+    let path = config.cwd().join(opts.path);
+    if fs::metadata(&path).is_ok() {
+        return Err(human(format!("Destination `{}` already exists",
+                                 path.display())))
+    }
+    
+    let name = try!(get_name(&path, &opts, config));
+    try!(check_name(name));
+
+    let mkopts = MkOptions {
+        version_control: opts.version_control,
+        path: &path,
+        name: name,
+        source_files: vec![plan_new_source_file(opts.bin, String::from(name))],
+    };
+    
+    mk(config, &mkopts).chain_error(|| {
+        human(format!("Failed to create project `{}` at `{}`",
+                      name, path.display()))
+    })
+}
+
+pub fn init(opts: NewOptions, config: &Config) -> CargoResult<()> {
+    let path = config.cwd().join(opts.path);
+    
+    let cargotoml_path = path.join("Cargo.toml");
+    if fs::metadata(&cargotoml_path).is_ok() {
+        return Err(human(format!("Destination `{}` already exists",
+                                 cargotoml_path.display())))
+    }
+    
+    let name = try!(get_name(&path, &opts, config));
+    try!(check_name(name));
+    
+    let mut src_paths_types = vec![];
+    
+    try!(detect_source_paths_and_types(&path, name, &mut src_paths_types));
+    
+    if src_paths_types.len() == 0 {
+        src_paths_types.push(plan_new_source_file(opts.bin, String::from(name)));
+    } else {
+        // --bin option may be ignored if lib.rs or src/lib.rs present
+        // Maybe when doing `cargo init --bin` inside a library project stub,
+        // user may mean "initialize for library, but also add binary target"
+    }
+    
+    let mut version_control = opts.version_control;
+    
+    if version_control == None {
+        let mut num_detected_vsces = 0;
+        
+        if fs::metadata(&path.join(".git")).is_ok() {
+            version_control = Some(VersionControl::Git);
+            num_detected_vsces += 1;
+        }
+        
+        if fs::metadata(&path.join(".hg")).is_ok() {
+            version_control = Some(VersionControl::Hg);
+            num_detected_vsces += 1;
+        }
+        
+        // if none exists, maybe create git, like in `cargo new`
+        
+        if num_detected_vsces > 1 {
+            return Err(human("Both .git and .hg exist. It is not recommended \
+                             to use multiple VCS'es simultaneously. \
+                             I don't know which one to choose to fill in the \
+                             respective ignore file. Specify --vcs option \
+                             to override this detection"));
```

Comment:
Error messages in Cargo tend to not take the first person point of view, perhaps something along the lines of:

```
both .git and .hg folders found and the ignore file can't be 
filled in as a result, specify --vcs to override detection
```